h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -2470,62 +2679,67 @@ public void onCompletion(CountedCompleter caller) {
         return !stop_requested() && _state._iter < _parms._max_iterations && !_earlyStop;
       } else {
         GLMGradientInfo gginfo = (GLMGradientInfo) ginfo;
-        if(gginfo._gradient == null)
-          _state.updateState(beta,gginfo._likelihood);
+        if (gginfo._gradient == null)
+          _state.updateState(beta, gginfo._likelihood);
         else
           _state.updateState(beta, gginfo);
-        if (!_parms._lambda_search)
+        if (!_parms._lambda_search || (_parms._generate_scoring_history && !_insideCVCheck))
           updateProgress(true);
         boolean converged = !_earlyStopEnabled && _state.converged(); // GLM specific early stop.  Disabled if early stop is enabled
         if (converged) Log.info(LogMsg(_state.convergenceMsg));
         return !stop_requested() && !converged && _state._iter < _parms._max_iterations && !_earlyStop;
       }
     }
 
-    public boolean progressHGLMGLMMME(double sumDiff2, double sumeta2, int iteration, boolean atGLMMME, GLMModel 
+    public boolean progressHGLMGLMMME(double sumDiff2, double sumeta2, int iteration, boolean atGLMMME, GLMModel
             fixedModel, GLMModel[] randModels, Frame glmmmeReturns, Frame hvDataOnly, double[] VC1, double[][] VC2,
                                       double[][] cholR, Frame augZ) {
-      boolean converged = !_earlyStopEnabled && (sumDiff2 < (_parms._objective_epsilon*sumeta2));
+      boolean converged = !_earlyStopEnabled && (sumDiff2 < (_parms._objective_epsilon * sumeta2));
       if (atGLMMME) {
         _state._iterHGLM_GLMMME++;
       } else {
-          _state._iter++;
-          updateProgress(fixedModel, randModels, glmmmeReturns, hvDataOnly, VC1, VC2, sumDiff2,
-                  sumDiff2 / sumeta2, true, cholR, augZ);
+        _state._iter++;
+        updateProgress(fixedModel, randModels, glmmmeReturns, hvDataOnly, VC1, VC2, sumDiff2,
+                sumDiff2 / sumeta2, true, cholR, augZ);
       }
       return !stop_requested() && !converged && (iteration < _parms._max_iterations) && !_earlyStop;
     }
-    
-    public boolean progress(double [] beta, double likelihood) {
+
+    public boolean progress(double[] beta, double likelihood) {
       _state._iter++;
-      _state.updateState(beta,likelihood);
-      if(!_parms._lambda_search)
+      _state.updateState(beta, likelihood);
+      if (!_parms._lambda_search || (_parms._generate_scoring_history && !_insideCVCheck))
```

Comment:
Do I read this correctly that when using lambda search, turning on `generate_scoring_history` should help with the early stopping (compared with the version before this PR)? (Assuming early stopping in lambda search uses `_earlyStop`)

Small experiment suggests that with `generate_scoring_history` and `score_each_iteration`, it updates early stop each iteration. If I turn off `generate_scoring_history` it doesn't.

Reply:
@tomasfryda  This part has nothing to do with early stopping.  It has something to do with cross-validation.  I added this comment to the code and added it here:

updateProgress is not run originally when lambda_search is not enabled inside or outside of 
cv_computeAndSetOptimalParameters.  However, with generate_scoring_history on, I am adding deviance_* and hence I need to make sure that updateProgress is not run inside cv_computeAndSetOptimalParameters.  Hence, the extra condition of (_parms._generate_scoring_history && !_insideCVCheck).  The insideCVCheck means inside cv_computeAndSetOptimalParameters.  If you dig into the cv_computeAndSetOptimalParameters code, it actually will call computeSubmodel(lidx, lambda, Double.NaN, Double.NaN) which will end up calling the progress(double[] beta, double likelihood) code which then calls updateProgress in line 2712 or 2713.

Reply:
@wendycwong  thank you for your explanation. My question regarding early stopping was motivated by
```python
glm = H2OGeneralizedLinearEstimator(
    generate_scoring_history=True,
    lambda_search=True,
    score_each_iteration=True,
    nfolds=0
 )
glm.train(y="CAPSULE", training_frame=df)
```
which calls `updateEarlyStop` every iteration but when I set `generate_scoring_history=False` then `updateEarlyStop` is not called at all. And when I look in `GLMDriver#doCompute`[1] it seems to me that by setting `generate_scoring_history` I could indirectly early stop lambda search but IIRC lambda search has its own early stopping mechanism so I'm asking in case there could be some conflict.

I'm mentioning it just because you mentioned CV and this should be a case without CV (`nfolds=0`).

[1] https://github.com/h2oai/h2o-3/blob/master/h2o-algos/src/main/java/hex/glm/GLM.java#L2379

Reply:
@tomasfryda That is a good point, I know _earlyStopEnable is set as

      if (!_parms._lambda_search && (_parms._stopping_rounds > 0))  // early stop is on!
        _earlyStopEnabled = true;

Hence, when lamda_search = true, _earlyStopEnabled will always be false.

I have changed:

_earlyStop = updateEarlyStop();

to 

_earlyStop = _earlyStopEnabled && updateEarlyStop(); 

This wil suppressed the call to updateEarlyStop() unnecessarily.  I don't really remember why there is _earlyStop and _earlyStopEnabled......